File: equivalences/equivalences.py
import sys
from pprint import pprint
sys.setrecursionlimit(200000)

class SCCGraph:
    def __init__(self, n_nodes):
        self.n_nodes = n_nodes
        self.i_edges = [[] for _ in range(n_nodes)]
        self.o_edges = [[] for _ in range(n_nodes)]
        self.visited = [False for _ in range(n_nodes)]
        self.visited_nodes = []
        self.components = {}
        self.component_assignments = [-1 for _ in range(n_nodes)]

    def add_edge(self, src, dst):
        self.o_edges[src].append(dst)
        self.i_edges[dst].append(src)

    def _visit(self, node):
        if not self.visited[node]:
            self.visited[node] = True
            for v in self.o_edges[node]:
                self._visit(v)
            self.visited_nodes.append(node)

    def _assign(self, node, root):
        if self.component_assignments[node] < 0:
            self.component_assignments[node] = root
            self.components.setdefault(root, []).append(node)
            for v in self.i_edges[node]:
                self._assign(v, root)

    def find_scc(self):
        for u in range(self.n_nodes):
            self._visit(u)
        # pprint(self.visited)
        # pprint(list(reversed(self.visited_nodes)))
        for u in reversed(self.visited_nodes):
            self._assign(u, u)


def main():
    n_cases = int(sys.stdin.readline().strip())
    for _ in range(n_cases):
        n_statements, n_proofs = (int(x) for x in sys.stdin.readline().split() if x)
        graph = SCCGraph(n_statements)
        for _ in range(n_proofs):
            src, dst = (int(x) - 1 for x in sys.stdin.readline().split() if x)
            graph.add_edge(src, dst)
        graph.find_scc()
        # pprint(graph.component_assignments)
        # pprint(graph.components)
        if len(graph.components) == 1:
            print(0)
        else:
            n_zero_i = 0
            n_zero_o = 0
            for root, nodes in graph.components.items():
                i_degree = sum(len([e for e in graph.i_edges[node] if graph.component_assignments[e] != root]) for node in nodes)
                o_degree = sum(len([e for e in graph.o_edges[node] if graph.component_assignments[e] != root]) for node in nodes)
                if i_degree == 0:
                    n_zero_i += 1
                if o_degree == 0:
                    n_zero_o += 1
            print(max(n_zero_i, n_zero_o))

main()

File: equivalences/equivalences.cpp
#define _USE_MATH_DEFINES
#include <iostream>
#include <cmath>
#include <cstdio>
#include <vector>
#include <map>

using namespace std;

typedef vector<vector<int> > vvi;

struct Graph {
    int n_nodes;
    vvi i_edges;
    vvi o_edges;
    vector<char> visited;
    vector<int> visited_nodes;
    map<int, vector<int> > components;
    vector<int> component_assignments;

    Graph(int n_nodes)
    {
        this->n_nodes = n_nodes;
        i_edges.resize(n_nodes);
        o_edges.resize(n_nodes);
        visited.resize(n_nodes);
        visited_nodes.reserve(n_nodes);
        component_assignments.resize(n_nodes, -1);
    }

    void add_edge(int src, int dst)
    {
        o_edges[src].push_back(dst);
        i_edges[dst].push_back(src);
    }

    void visit(int node)
    {
        if (!visited[node])
        {
            visited[node] = 1;
            for (int v: o_edges[node])
                visit(v);
            visited_nodes.push_back(node);
        }
    }

    void assign(int node, int root)
    {
        if (component_assignments[node] < 0)
        {
            component_assignments[node] = root;
            components[root].push_back(node);
            for (int v: i_edges[node])
                assign(v, root);
        }
    }

    void find_scc()
    {
        for (int u = 0; u < n_nodes; ++u)
            visit(u);
        for (auto uit = visited_nodes.rbegin(); uit != visited_nodes.rend(); ++uit)
            assign(*uit, *uit);
    }
};

int main()
{
    int n_cases;
    cin >> n_cases;
    for (int ii = 0; ii < n_cases; ++ii)
    {
        int n_nodes, n_edges;
        cin >> n_nodes >> n_edges;
        Graph graph(n_nodes);
        for (int jj = 0; jj < n_edges; ++jj)
        {
            int src, dst;
            cin >> src >> dst;
            graph.add_edge(src-1, dst-1);
        }
        graph.find_scc();
        if (graph.components.size() == 1)
        {
            cout << 0 << endl;
        }
        else
        {
            int n_zero_o = 0;
            int n_zero_i = 0;
            for (auto it = graph.components.begin(); it != graph.components.end(); ++it)
            {
                int i_degree = 0;
                int o_degree = 0;
                for (int node: it->second)
                {
                    for (int src: graph.i_edges[node])
                        if (graph.component_assignments[src] != it->first)
                            i_degree++;
                    for (int dst: graph.o_edges[node])
                        if (graph.component_assignments[dst] != it->first)
                            o_degree++;
                }
                if (i_degree == 0)
                    n_zero_i += 1;
                if (o_degree == 0)
                    n_zero_o += 1;
            }
            cout << (max(n_zero_i, n_zero_o)) << endl;
        }
    }
    return 0;
}
